### ISSUE TITLE
Virtual method return types static and $this don't work

### ISSUE THREAD

**Issue by vlada-dudr** [phpactor/phpactor#1802]
```php
<?php

/**
 * @method static bla()
 */
class h {
}

$a = new h;
$b = $a->bla();
```
Expected result: `bla` method returns h
Actual: Method "bla" does not exist on class "h"

Changing `static` to `$this` is same. `self` makes phpactor discover the function, but doesn't mean the same thing...

**Comment by ging-dev** [comment]
@vlada-dudr try `@method static void bla()`

**Comment by vlada-dudr** [comment]
Well, that results in, as expected, Phpactor picking up static method on class h. But I want to describe instance method returning `static` type i.e. the class bla() was called on, as in `new static()`. If `$this` would work, it would cover most of the use cases.

**Comment by ging-dev** [comment]
> Well, that results in, as expected, Phpactor picking up static method on class h. But I want to describe instance method returning `static` type i.e. the class bla() was called on, as in `new static()`. If `$this` would work, it would cover most of the use cases.

Did you mean
```php
/** @method User get() */
class h {
}

$a = new h;
$b = $a->get() // $b is User
```
Or
```php
/** @method self get() */
class h {}

$a = new h;
$b = $a->get() // $b is h
```

**Comment by vlada-dudr** [comment]
`static` and `$this` are almost same, except that `$this` requires return of same instance. Compare:
```php
class a {
    public function x(): self { return new self(); }
    public function y(): static {return new static(); }
}

class b extends a {
}

$o = new b();
assert(get_class($o->x) === 'a');
assert(get_class($o->y) === 'b');
```
So there are two meanings `static` has in `@method` context. A property of a method or a return type. So, I expect that phpactor should distinguish between them. This should be legal: `@method static static make()`. Which describes static function which return instance of class it was called on as opposed to `@method static self make()`, which is static method that should return instance of class the method is defined on. In case of `@method static setWhatever()` the return type should be `static` - the class the method was called on, not a static method of such class.

Is it more understandable now?

The other part of the problem is that line `@method $this setSomething()` is ignored by phpactor, so none of them are usable.

**Comment by vlada-dudr** [comment]
I think it used to work some time ago. Maybe addition of virtual static methods did break parsing of static type? Because when I add `@method static bla()` I see nameless static method in the autocomplete suggestions.

**Comment by ging-dev** [comment]
> I think it used to work some time ago. Maybe addition of virtual static methods did break parsing of static type? Because when I add `@method static bla()` I see nameless static method in the autocomplete suggestions.

i think your usage is invalid because Psalm also don't understand it

https://psalm.dev/r/7bbbc5a0fc

**Comment by vlada-dudr** [comment]
`@method static static bla()` somewhat works, but is resolved like `@method static self bla()`.

**Comment by vlada-dudr** [comment]
@ging-dev your Psalm example is not what I mean.  `@method` annotations seem to work on that page only if you tick a "Use PHPDoc methods and properties without magic call." checkbox in settings.

**Comment by vlada-dudr** [comment]
@dantleech part of the problem I tried to solve in #1806. But the to me it seems that `self` and `static` are resolved at the moment of parsing the docblock, so child classes inherit already resolved type. So `static` and `$this` types are resolved to `self`.

The docblock parser always grabs `static` without context, so `static` type can be parsed only after `static` token.  I guess there will be some bactracking needed. But my knowlegde about phpactor base is too small, to dare to implement fix for this.

**Comment by dantleech** [comment]
The key part is in the TypeConverter:

https://github.com/phpactor/phpactor/blob/0ccdbaaec10ec3aabfaf1cf1e2209b2810069391/lib/WorseReflection/Bridge/Phpactor/DocblockParser/TypeConverter.php#L323-L326

The `resolver` is swapped out depending on the context, so for a ReflectionMethod it will be:

https://github.com/phpactor/phpactor/blob/0ccdbaaec10ec3aabfaf1cf1e2209b2810069391/lib/WorseReflection/Core/TypeResolver/MemberTypeResolver.php#L20-L30

And it will be resolved as the `Static` type f.e. not sure if we swap out for the "virtual" methods however, maybe an easy fix.

**Comment by vlada-dudr** [comment]
I have three tests to share:

**Comment by dantleech** [comment]
can you add them to your PR? or?

**Comment by vlada-dudr** [comment]
I can, but they are mostly failling. This issue is actually described by these three tests. My PR fixes only one of the failing asserts

**Comment by dantleech** [comment]
Coud be fixed via. https://github.com/phpactor/phpactor/issues/1823

**Comment by dantleech** [comment]
Fixed in #1855
